addons/rating/tests/test_controller.py

```diff
@@ -1,25 +1,69 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 from odoo.tests import HttpCase, tagged, new_test_user
+from odoo.addons.base.tests.common import HttpCaseWithUserDemo, HttpCaseWithUserPortal
 
 @tagged('post_install', '-at_install')
-class TestControllersRoute(HttpCase):
+class TestControllersRoute(HttpCaseWithUserDemo, HttpCaseWithUserPortal):
 
     def setUp(self):
         super(TestControllersRoute, self).setUp()
         self.user = new_test_user(self.env, "test_user_1", email="test_user_1@nowhere.com", tz="UTC")
         self.partner = self.user.partner_id
+        self.rated_partner = self.env['res.partner'].create({
+            'name': 'Test company',
+            'email': "testcompany@example.com",
+            'is_company': True,
+            'child_ids': [
+                (0, 0, {'name': 'Test child_1', 'type': 'contact', 'email': "testchild_1@example.com",}),
+            ]
+        })
 
     def test_controller_rating(self):
-
-        rating_test = self.env['rating.rating'].with_user(self.user).create({
-            'res_model_id': self.env['ir.model'].sudo().search([('model', '=', 'res.partner')], limit=1).id,
+        partner_model_id = self.env['ir.model'].sudo().search([('model', '=', 'res.partner')], limit=1).id
+        [rating_test_1, rating_test_2, rating_test_3] = self.env['rating.rating'].with_user(self.user).create([{
+            'res_model_id': partner_model_id,
             'res_model': 'res.partner',
-            'res_id': self.partner.id,
+            'res_id': self.rated_partner.id,
+            'partner_id': self.partner.id,
             'rating': 3
-        })
+        } for i in range(3)])
+
         self.authenticate(None, None)
-        access_token = rating_test.access_token
+        access_token = rating_test_1.access_token
         url = '/rate/%s/submit_feedback' % access_token
         req = self.url_open(url)
         self.assertEqual(req.status_code, 200, "Response should = OK")
+
+        # Test that internal user cannot give the rating
```

Comment:
All those tests could probably be run into a loop using subTest (give an expected output / expected code / expected string), allowing to reduce the test (and make it easier to read and understand various test cases)